File: python/second/three_sum.py
class Solution(object):
    # O(n^2) time | O(n) space
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        nums.sort()
        res = []

        for i in range(len(nums)):
            if nums[i] > 0:
                break
            if i > 0 and nums[i] == nums[i - 1]:
                continue
            num1 = nums[i]
            j = i + 1
            k = len(nums) - 1
            while j < k:
                num2 = nums[j]
                num3 = nums[k]
                sum = num1 + num2 + num3
                if sum == 0:
                    res.append([num1, num2, num3])
                    j += 1
                    k -= 1
                    while j < k and nums[j] == nums[j - 1]:
                        j += 1
                elif sum > 0:
                    k -= 1
                else:
                    j += 1
        return res
    
solution = Solution()
nums = [-1,0,1,2,-1,-4]
print(solution.threeSum(nums))

File: src/main/java/january_2026/ThreeSum.java
package january_2026;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class ThreeSum {

    public static void main(String[] args) {
//        int[] nums = {-1,0,1,2,-1,-4};
        int[] nums = {2,-3,0,-2,-5,-5,-4,1,2,-2,2,0,2,-4,5,5,-10};

        ThreeSum threeSum = new ThreeSum();
        List<List<Integer>> result = threeSum.threeSum(nums);
        System.out.println(result);
    }

    // O(n^2) time | O(n) space
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        Arrays.sort(nums);

        for (int i = 0; i < nums.length - 2; i++) {
            if (i >= 1 && nums[i - 1] == nums[i]) {
                continue;
            }
            int j = i + 1;
            int k = nums.length - 1;
            int num1 = nums[i];

            while (j < k) {
                int num2 = nums[j];
                int num3 = nums[k];

                int sum = num1 + num2 + num3;
                if (sum == 0) {
                    result.add(List.of(num1, num2, num3));
                    j++;
                    k--;
                    while (j < k && nums[j] == nums[j - 1]) {
                        j++;
                    }
                } else if (sum < 0) {
                    j++;
                } else {
                    k--;
                }
            }
        }

        return result;
    }

}
